src/libraries/System.Net.WebSockets.Client/src/System/Net/WebSockets/BrowserWebSockets/BrowserWebSocket.cs

```diff
@@ -471,160 +375,247 @@ private async Task ConnectAsyncCore(CancellationToken cancellationToken)
+                await CancellationHelper(closeTask, cancellationToken, previousState).ConfigureAwait(false);
             }
-#endif
-
-#if FEATURE_WASM_THREADS
-            lock (_thisLock)
+            if (waitForCloseReceived)
             {
-#endif
-                var state = State;
-                if (state == WebSocketState.Open || state == WebSocketState.Connecting || state == WebSocketState.CloseSent)
+                lock (_lockObject)
                 {
-                    FastState = waitForCloseReceived ? WebSocketState.Closed : WebSocketState.CloseSent;
+                    _closeReceived = true;
+                    ForceReadCloseStatus();
+                    _ = State;
                 }
-#if FEATURE_WASM_THREADS
-            } //lock
-#endif
+            }
         }
 
-        private async Task CancelationHelper(Task jsTask, CancellationToken cancellationToken, WebSocketState previousState)
+        private async Task CancellationHelper(Task promise, CancellationToken cancellationToken, WebSocketState previousState, IDisposable? disposable = null)
         {
-            if (jsTask.IsCompletedSuccessfully)
+            if (cancellationToken.IsCancellationRequested)
             {
+                disposable?.Dispose();
+                cancellationToken.ThrowIfCancellationRequested();
+            }
+            if (promise.IsCompletedSuccessfully)
+            {
+                disposable?.Dispose();
                 return;
             }
             try
             {
                 using (var receiveRegistration = cancellationToken.Register(static s =>
                 {
                     CancelablePromise.CancelPromise((Task)s!);
-                }, jsTask))
+                }, promise))
                 {
-                    await jsTask.ConfigureAwait(true);
+                    await promise.ConfigureAwait(false);
                     return;
                 }
             }
             catch (JSException ex)
             {
-                if (State == WebSocketState.Aborted)
+                lock (_lockObject)
                 {
-                    throw new OperationCanceledException(nameof(WebSocketState.Aborted), ex);
+                    var state = State;
+                    if (state == WebSocketState.Aborted)
+                    {
+                        ForceReadCloseStatus();
+                        throw new OperationCanceledException(nameof(WebSocketState.Aborted), ex);
+                    }
+
+                    if (state != WebSocketState.Closed && cancellationToken.IsCancellationRequested)
+                    {
+                        FastState = WebSocketState.Aborted;
+                        _cancelled = true;
+                        throw new OperationCanceledException(cancellationToken);
+                    }
+                    if (state != WebSocketState.Closed && ex.Message == "Error: OperationCanceledException")
```

Comment:
Can we do anything better than checking exception message (storing a flow on the exception or anything?

Reply:
this is rather fallback in case `cancellationToken.IsCancellationRequested` didn't catch it. It's not worth the trouble to complicate it, we own the message so it would not change on it's own.

Reply:
I don't thing it will be obvious when we do localization for example, but it's your call